### ISSUE TITLE
Inline completions - user feedback 

### ISSUE THREAD

**Issue by dimacodota** [codota/TabNine#432]
Hi All, we just released an experimental "inline Completions" feature 🎉 

Please share your experience with this feature🙏 

In order to enable it: 
 - make sure you are using the latest engine(3.5.37) and extension(3.4.21) version.
 - open the Tabnine  Home(from the status bar item), scroll to "General Preferences" and check the "inline Completions"  option
 - restart vscode
 
<img width="1246" alt="Screen Shot 2021-08-08 at 11 15 14" src="https://user-images.githubusercontent.com/60742964/128625904-da021af1-dbd9-48f8-907e-d2713486ea36.png">

**Comment by rcoundon** [comment]
I'm possibly being dumb here but where is Tabnine Home? I can only see the regular VSCode preferences and I don't seem to be able to find anywhere to download the Tabnine engine separately (i.e. only the various extensions) which is what's implied by the instructions.

**Comment by dimacodota** [comment]
> I'm possibly being dumb here but where is Tabnine Home? I can only see the regular VSCode preferences and I don't seem to be able to find anywhere to download the Tabnine engine separately (i.e. only the various extensions) which is what's implied by the instructions.

click on the "Tabnine" status bar item:
<img width="491" alt="Screen Shot 2021-08-08 at 12 07 15" src="https://user-images.githubusercontent.com/60742964/128626968-40bee91a-59ac-475e-901d-2b74f809b0c7.png">

**Comment by rcoundon** [comment]
> > I'm possibly being dumb here but where is Tabnine Home? I can only see the regular VSCode preferences and I don't seem to be able to find anywhere to download the Tabnine engine separately (i.e. only the various extensions) which is what's implied by the instructions.
> 
> click on the "Tabnine" status bar item:
> <img alt="Screen Shot 2021-08-08 at 12 07 15" width="491" src="https://user-images.githubusercontent.com/60742964/128626968-40bee91a-59ac-475e-901d-2b74f809b0c7.png">

Great - thank you

**Comment by dreambit** [comment]
Does it allow to see one suggestion only?

**Comment by dimacodota** [comment]
> Does it allow to see one suggestion only?
you can "list" through the suggestions via `alt +[` and `alt +]` keys
<img width="648" alt="Screen Shot 2021-08-12 at 0 53 11" src="https://user-images.githubusercontent.com/60742964/129108411-a773204c-2c41-468d-92a6-f1b8af816ba4.png">

**Comment by StefnirKristjansson** [comment]
I honestly find it better than the popup since I have a lot of snippets that I use the tab space I didn't like the fact that the suggestions crowded out the snippets.

What would be perfect is that I could see the other suggestions somewhere so it don't waste time I circling them with the alt +[

Since I used tabnine to be faster I would like to see it fast if there are any suggestions that I can use.

**Comment by Xusine1131** [comment]
When using lens-related plugins like Error lens and Git lens, it breaks Tabnine's autocomplete.

**Comment by HippieHosenHustler** [comment]
Looks like a cool future. It does not seem to work in the IntelliJ Plugin though. Is this something that is in the works?

**Comment by Eransho** [comment]
> Looks like a cool feature. It does not seem to work in the IntelliJ Plugin though. Is this something that is in the works?
Yes, we are working on it. It should be released soon

**Comment by DenysAshikhin** [comment]
Love this feature when combined with the Full Line completions. However, I have two requests/questions:

1) Would it be possible to add a switch (or be default) that if we have full line completion turned on, it will instead show its longest prediction first. I just find myself sometimes having to use the 'ctr  + [ OR ]' to cycle from completing the current word (which is accurate) to the full line which is a bit cumbersome

2) Somehow combine inline with a dropdown? Like it would show the first auto-complete suggestion inline, but all other options would appear in a dropdown (either the current drop down style, or a different one).

#2 is key because it shows whether I will waste time cyclying through each one with the hot keys or not.

Other than that, great work!

**Comment by IRediTOTO** [comment]
love this feature :)

**Comment by laundmo** [comment]
In VSCode the sugestion sometimes appears in front of the text. Seems to only happen when starting a new line.
![](https://img.laundmo.com/u/748HJM.png)

**Comment by shmkane** [comment]
There's occasions where pressing Tab to accept an inline-completion results in either an actual tab being inserted, or it prefers the first option of a dropdown Intellisense suggestion instead. 

To prevent the latter, I've completely disabled the dropdowns from VSCode entirely. 

However, in this .md file for example, the issue still persists 

https://user-images.githubusercontent.com/2586027/130464528-21d5733a-ab00-4035-a3e1-a0d86136a429.mov

Settings: 
![image](https://user-images.githubusercontent.com/2586027/130466188-f19ce5c9-4e64-48ec-9873-8d615fc8b824.png)
![image](https://user-images.githubusercontent.com/2586027/130466229-3abb6b08-b28f-468f-8e09-6e52cafba3e7.png)

**Comment by dimacodota** [comment]
> There's occasions where pressing Tab to accept an inline-completion results in either an actual tab being inserted, or it prefers the first option of a dropdown Intellisense suggestion instead.
> 
> To prevent the latter, I've completely disabled the dropdowns from VSCode entirely.
> 
> However, in this .md file for example, the issue still persists
> 
>  Screen.Recording.2021-08-23.at.10.26.24.AM.mov 
> Settings:
> ![image](https://user-images.githubusercontent.com/2586027/130466188-f19ce5c9-4e64-48ec-9873-8d615fc8b824.png)
> ![image](https://user-images.githubusercontent.com/2586027/130466229-3abb6b08-b28f-468f-8e09-6e52cafba3e7.png)

@shmkane, can you please share your console log (Help -> Toggle Developer Tools) ?

**Comment by shmkane** [comment]
> > There's occasions where pressing Tab to accept an inline-completion results in either an actual tab being inserted, or it prefers the first option of a dropdown Intellisense suggestion instead.
> > To prevent the latter, I've completely disabled the dropdowns from VSCode entirely.
> > However, in this .md file for example, the issue still persists
> > Screen.Recording.2021-08-23.at.10.26.24.AM.mov
> > Settings:
> > ![image](https://user-images.githubusercontent.com/2586027/130466188-f19ce5c9-4e64-48ec-9873-8d615fc8b824.png)
> > ![image](https://user-images.githubusercontent.com/2586027/130466229-3abb6b08-b28f-468f-8e09-6e52cafba3e7.png)
> 
> @shmkane, can you please share your console log (Help -> Toggle Developer Tools) ?


Here's me attempting to tab complete on a markdown file vs a javascript file.


https://user-images.githubusercontent.com/2586027/130484757-fd01a9c4-f04d-4802-b0c2-0873b9044a5a.mov


I'm unsure if the errors displayed are related, but here's a [PasteBin](https://pastebin.com/nZh6ypNW) of the output shown in the video

**Comment by dimacodota** [comment]
@shmkane - can you please share the  list of your installed extensions `code --list-extensions`?

**Comment by shmkane** [comment]
> @shmkane - can you please share the list of your installed extensions `code --list-extensions`?

`code --list-extensions`
```
christian-kohler.npm-intellisense
christian-kohler.path-intellisense
CoenraadS.bracket-pair-colorizer-2
dbaeumer.vscode-eslint
discountry.react-redux-react-router-snippets
donjayamanne.python-extension-pack
eamodio.gitlens
eg2.vscode-npm-script
esbenp.prettier-vscode
janisdd.vscode-edit-csv
jasonnutter.search-node-modules
jawandarajbir.react-vscode-extension-pack
k--kato.intellij-idea-keybindings
KevinRose.vsc-python-indent
magicstack.MagicPython
ms-python.python
ms-python.vscode-pylance
naumovs.color-highlight
njpwerner.autodocstring
richie5um2.vscode-sort-json
rvest.vs-code-prettier-eslint
streetsidesoftware.code-spell-checker
TabNine.tabnine-vscode
VisualStudioExptTeam.vscodeintellicode
vscode-icons-team.vscode-icons
xabikos.JavaScriptSnippets
xabikos.ReactSnippets
yzhang.markdown-all-in-one
zhuangtongfa.material-theme
```

**Comment by xuzhanhh** [comment]
it seems quit edit mode and turn visual mode with vsocde vim in some cases

**Comment by Magnusri** [comment]
Fantastic feature!

I think it would be great if pressing arrow down or up(to select from the regular dropdown list) would disable the feature temporarily to allow tab to pick from the regular dropdown instead.

**Comment by hugohabicht01** [comment]
What I personally noticed is that the inline completetions aren't really compatible with the vscode vim plugin (`vscodevim.vim`). When using a suggested completion by pressing tab, the vim plugin will go into visual mode, the vim cursor stays at the last typed character but a "normal", thin cursor will appear at the end of the completed code. This is quite an annoyance as you can't keep typing but instead need to spam esc to get back into normal mode and then move your cursor to the end again and go into insert mode before you can start typing again.

**Comment by shadowndacorner** [comment]
I've experienced the same issues as several others in the thread with this feature with the Vim extension for VS Code.

I also noticed similar issues with the markdown extension breaking tab auto completion as described by @shmkane.  Based on my own experience with a simple navigation extension I wrote, it seems like the markdown extension specifically captures input overzealously - I noticed that it broke the backspace keybinding I use in my extension.  I've only noticed that behavior when editing markdown files so far.  Not sure if that's helpful at all for narrowing down that issue.

As a feature request, it would be nice if there was a way to request inline suggestions, similar to how `ctrl+space` opens the regular suggestion dropdown.

**Comment by frizzant** [comment]
I sometimes just use Tabnine as a kind of AI-assisted lexicon of what i may need, so I would suggest that it does show the other options too, instead of having to manually expand the other options, I'd like to see the inline suggestion, and the others too. 
(PHPStorm)

**Comment by jrrmcalcio** [comment]
I can't make it work, when a suggestion appears, i hit tab and just a TAB character appears :( I'm using VSCode 1.59.1, tabnine plugin: 3.4.26 and engine: 3.6.5-40 pro.  Any help?

Thank you

**Comment by igorsgm** [comment]
Does this feature work with Jetbrains at all? I can't make it work on PHPStorm.

**Comment by IRediTOTO** [comment]
VS code just release new version inludes inline complete. Should I remove tab9 ? So confuse

**Comment by IRediTOTO** [comment]
I feel tab9 still smarter now

**Comment by Eransho** [comment]
> Does this feature work with Jetbrains at all? I can't make it work on PHPStorm.

@igorsgm it should work with JetBrains IDEs. Is the inline completion option in your Tabnine hub is on? What is your tabnine plugin version?

**Comment by hauselin** [comment]
> VS code just release new version inludes inline complete. Should I remove tab9 ? So confuse

I have the same question too. How does this overlap with vscode's [suggest preview feature](https://code.visualstudio.com/updates/v1_60#_inline-suggestions-in-autocomplete)?

Also, this feature doesn't work very well with [vscode vim](https://github.com/VSCodeVim/Vim).

**Comment by Joeri-KIXX** [comment]
The inline completions is totally unusable for me (working in Laravel 8)
The code randomly disappears and is all messed up.

I made a video to show you what happens:
https://app.vmaker.com/record/1BlZnP8UYeeeZe3g

**Comment by CharlonTank** [comment]
Sometime, it create newlines and remove some text, I guess it's because the indentations is wrong.

Mac OS 11.5
VS Code: 1.61.0-insider (Universal)
Language: elm
Extension: Pro version - Last updated 01/09/2021, 13:14:46

**Comment by bigntallmike** [comment]
I really want this feature to work, but while editing Python files in VS Code, I find it will suddenly start entering large amounts of text where I don't want it and there's just too much jumping around visually to track.

**Comment by hienlh** [comment]
Something went wrong when Im using this feature.
![CleanShot 2021-09-10 at 09 27 54](https://user-images.githubusercontent.com/36977998/132789549-b770680d-0bbf-40c5-8d4f-16607f6ffefc.gif)

**Comment by CharlonTank** [comment]
> Something went wrong when Im using this feature.
> 
> ![CleanShot 2021-09-10 at 09 27 54](https://user-images.githubusercontent.com/36977998/132789549-b770680d-0bbf-40c5-8d4f-16607f6ffefc.gif)
> 
> I think the issue happened when combining with this feature
> 
> ![CleanShot 2021-09-10 at 09 30 29](https://user-images.githubusercontent.com/36977998/132789725-37913911-4cff-43cf-bf70-c0dc6b06b82f.png)

I get this problem as well ^

**Comment by nosolosoft** [comment]
> > Something went wrong when Im using this feature.
> > ![CleanShot 2021-09-10 at 09 27 54](https://user-images.githubusercontent.com/36977998/132789549-b770680d-0bbf-40c5-8d4f-16607f6ffefc.gif)
> > I think the issue happened when combining with this feature
> > ![CleanShot 2021-09-10 at 09 30 29](https://user-images.githubusercontent.com/36977998/132789725-37913911-4cff-43cf-bf70-c0dc6b06b82f.png)
> 
> I get this problem as well ^

Something similar here!

**Comment by jarenal** [comment]
I love this feature, the only problem I find is that when I accept a suggestion by pressing TAB, it leaves the CURSOR at the end of the sentence, and it would be ideal if the accepted suggestion is a function or method, leave the CURSOR inside the parentheses to start to write directly the arguments of the function or method. Thank you!

Current behavior:

`$this->fooMethod()[CURSOR]`

Desired behavior:

`$this->fooMethod( [CURSOR] )`

**Comment by AliXEagle** [comment]
> it seems will quit edit mode and turn to visual mode with vsocde vim in some cases

Same here

**Comment by Vaillus** [comment]
> > > Something went wrong when Im using this feature.
> > > ![CleanShot 2021-09-10 at 09 27 54](https://user-images.githubusercontent.com/36977998/132789549-b770680d-0bbf-40c5-8d4f-16607f6ffefc.gif)
> > > I think the issue happened when combining with this feature
> > > ![CleanShot 2021-09-10 at 09 30 29](https://user-images.githubusercontent.com/36977998/132789725-37913911-4cff-43cf-bf70-c0dc6b06b82f.png)
> > 
> > 
> > I get this problem as well ^
> 
> Something similar here!

Same for me!
But it deletes my line of code as well. Pretty annoying, hope you'll find a way to fix it :)

**Comment by shadowndacorner** [comment]
As an FYI for those who are having issues with markdown (@shmkane for one), the markdown extension has a default keybinding for "tab".  Removing that keybinding allowed me to use this correctly.

@dimacodota As for the Vim issues, it seems like [this issue](https://github.com/Microsoft/vscode/issues/13441) may be the cause, but that definitely isn't certain.  That may just be an issue of a conflicting keybinding as well.

**Comment by arsdragonfly** [comment]
> > > > Something went wrong when Im using this feature.
> > > > ![CleanShot 2021-09-10 at 09 27 54](https://user-images.githubusercontent.com/36977998/132789549-b770680d-0bbf-40c5-8d4f-16607f6ffefc.gif)
> > > > I think the issue happened when combining with this feature
> > > > ![CleanShot 2021-09-10 at 09 30 29](https://user-images.githubusercontent.com/36977998/132789725-37913911-4cff-43cf-bf70-c0dc6b06b82f.png)
> > > 
> > > 
> > > I get this problem as well ^
> > 
> > 
> > Something similar here!
> 
> Same for me! But it deletes my line of code as well. Pretty annoying, hope you'll find a way to fix it :)

I'm having similar issues and I'm using vim mode. Really hope that it gets fixed.

**Comment by agi1512** [comment]
Hi. As of my experience (almost a week), i think it requires a minor improvement. Sometimes i type too fast, but ping to the server (i use cloud completions) is faster than my reaction time. So the problem is, when my reaction time is not good enough i type exactly as the model predicted for a couple of characters forward and i want to press tab to complete, it already changed it's prediction to something else (usually less useful), requiring me to spend more time to undo changes and redo it.

Also little about multiline suggestions in inline mode. They are good (in a sense), but don`t work properly on pycharm, firstly they dont print \n character at all just some stuff like `a=1print(a)` instead of something like this
 ```python
a=1 
print(a)
```
or at least like `a=1 \n print(a)`
And also they use tabs instead of spaces, which makes IDE panic. Looking forward for more stable inline mode, but for now switching to standard mode.

**Comment by ThEndGuy** [comment]
For everyone having trouble with [vscodevim](https://marketplace.visualstudio.com/items?itemName=vscodevim.vim) entering visual mode when using this feature, I found a workaround:

1. Install [macros](https://marketplace.visualstudio.com/items?itemName=geddski.macros) extension;
2. Add this into your settings.json:
`
"macros":
{
        "vim&tab9Insert": [
            "tabnine.accept-inline-suggestion",
            "extension.vim_escape",
            {
                "command": "type",
                "args": {
                    "text": "a"
                }
            },
        ]
    },`

3.   Add this into your keybindings.json (change the 'tab' key to whatever you want):
   `{
        "key": "tab",
        "command": "macros.vim&tab9Insert",
        "when": "tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled && vim.mode == 'Insert' || tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled && vim.mode == 'Insert'"
    },
`

Hopefully it should just work after this.

**Comment by Foxhoundn** [comment]
My 2 cents here as well - it works somewhat OK except when you decide to not use TabNine's suggestion and start cycling through VSCode's suggestions list - at that moment TabNine should be out of the picture.

Instead what happens is that I select whatever I want from the suggestion list, press [Tab] and of course TN inline prediction gets added instead :)

**Comment by sebastianjung** [comment]
vscode vim breaks when typing one letter and autocomplete through tabnine. When empty line competions is enabled the autocomplete will work before typing any letter.

**Comment by FyxedHasGithub** [comment]
> > Does it allow to see one suggestion only?
> > you can "list" through the suggestions via `alt +[` and `alt +]` keys
> 
> <img alt="Screen Shot 2021-08-12 at 0 53 11" width="648" src="https://user-images.githubusercontent.com/60742964/129108411-a773204c-2c41-468d-92a6-f1b8af816ba4.png">

i dont get the bar up, but i can use those commands, how can i make that bar show "the escape and enter" and how can i accept inline suggestions using enter key

**Comment by will961123** [comment]
How to modify the default completion shortcut key? Because she conflicts with Emmet, my IDE is webstorm

**Comment by sahuyash1103** [comment]
I have a problem with inline completion.
When I try to indent a line, it auto completes, and it irritates me a lot.
Can anyone from tabenine do something like disabling the tab key for autocompletion or another key for auto-completion.
It will make this issue solve.

**Comment by sahuyash1103** [comment]
> 

Can we set it to double-space for autocompletion and tab for regular indentation?

**Comment by yonip23** [comment]
Hi all,
If you're using JetBrains and working with inline completions, please provide your feedback here: https://github.com/codota/tabnine-intellij/issues/257

Thanks!

**Comment by elsandosgrande** [comment]
It does not seem possible to cycle through inline suggestions with at least some non-English keyboard layouts, at least in Visual Studio Code, because the AltGr key needed to type `[` and `]` is just registered as an Alt key, probably because that's where the right Alt key is in keyboard layouts without an AltGr key. I'd show a comparison between the English UK and Bosnian layouts to drive this point home, but both layouts have an AltGr key on Linux because both layouts have more characters available in them than their Windows and (presumably) macOS counterparts.

tl;dr: AltGr is needed to type `[` and `]`, but it's equal to Alt to VS Code, so no inline suggestion cycling for Bosnians for example.

**Comment by laundmo** [comment]
> It does not seem possible to cycle through inline suggestions with at least some non-English keyboard layouts, at least in Visual Studio Code, because the AltGr key needed to type `[` and `]` is just registered as an Alt key, probably because that's where the right Alt key is in keyboard layouts without an AltGr key. I'd show a comparison between the English UK and Bosnian layouts to drive this point home, but both layouts have an AltGr key on Linux because both layouts have more characters available in them than their Windows and (presumably) macOS counterparts.
> 
> tl;dr: AltGr is needed to type `[` and `]`, but it's equal to Alt to VS Code, so no inline suggestion cycling for Bosnians for example.

till this i wasnt even aware cycling through inline suggestions was possible

definitely the same issue on german keyboards tho

**Comment by STeeL835** [comment]
I'd like to find a way to change the color of the autocompletion token - it's brighter than a comment token and is indistinguishable from regular text on my theme (Kay Light) - confuses me, especially when I edit something in the middle of the line.
![image](https://user-images.githubusercontent.com/25896927/150136664-56a74ea1-9f46-46ec-8c7f-c86a2ee40ada.png)

**Comment by yonip23** [comment]
@STeeL835 Currently there's no such option, the plugin tries to find the "best" shade of gray per theme and I presume it doesn't work perfectly... 
Your point is valid - I'll discuss it with the dev team

Thanks for the feedback!

**Comment by StefanoV1989** [comment]
Nice! I love the inline completions because in the popup i prefer the real api reference.
For example if i need to call $api->method, the tabnine suggestions sometime don't allow me to see the real method name, and it tries to autocompile with used words. With inline mode i can use "enter" for popup, and "tab" to confirm for tabnine suggestion.

**Comment by yonip23** [comment]
@STeeL835 Check out the latest Tabnine plugin version for Intellij - 0.5.15 🥳 
We added a settings panel, where you can configure the hint color:
![image](https://user-images.githubusercontent.com/67855609/150670028-0807cefb-9e89-426c-9177-9240e8f32450.png)

**Comment by STeeL835** [comment]
@yonip23, that's cool! However, I should've mentioned this in my original post, but I'm using VS Code. Sorry

**Comment by yonip23** [comment]
> @yonip23, that's cool! However, I should've mentioned this in my original post, but I'm using VS Code. Sorry

🤣 Well, so at least you helped some JB users enjoy the plugin a bit more, We'll see if such feature can be added to VSCode as well

**Comment by yonip23** [comment]
@STeeL835 Unfortunately we can't provide such feature in the long term, due to some major changes that we plan on doing in the inline api in the next few weeks. But for now, I added a config property available in settings -> Tabnine -> Inline Hint Color.
Enjoy it while you can 🙂
Available in version 3.5.17 
![image](https://user-images.githubusercontent.com/67855609/150749289-d4fbba88-788e-4fbe-9e6e-daeb00c6377a.png)

**Comment by Roboroads** [comment]
As requested - here are my findings: I turned it off. This is because: 
 
- It is often far too late. \
I type `this.get();` and then recommends me `();` to put behind it. Yes - it would have been a good suggestion, it's just too late.
- It also does not really look ahead of the cursor, only whats behind. \
I have a code `this.get({options: true});` and decide to change `true` to `false` by removing just `true` and start typing `false`. After that, it recommends me to do `});` after, but that's already there - so acceptiong it would end up in `this.get({options: false});});`

It was just annoying enough and not helpful enough to leave on. Other than that this one has a lot of potential!

**Comment by yonip23** [comment]
@Roboroads thanks for the feedback! What plugin version are you on?

**Comment by Roboroads** [comment]
> @Roboroads thanks for the feedback! What plugin version are you on?

JetBrains plugin 0.5.15

**Comment by yonip23** [comment]
> JetBrains plugin 0.5.15

Yeah, that's the latest one currently. Important feedback, I'll discuss with my team. 
Thanks!

**Comment by IsaacShelton** [comment]
Cool feature, but it seems to break when using with VSCodeVim. Pressing `tab` to accept a completion does the completion successfully, but the caret in split in two, with one visual mode cursor and one weird cursor that isn't in any mode.

Have no idea what why

---

Typing code...

_single cursor (insert mode)_
<img width="270" alt="Screen Shot 2022-02-01 at 6 48 04 PM" src="https://user-images.githubusercontent.com/14429506/152075578-5c9c5818-a600-4a19-9b6b-f7ec5283d20c.png">

After `tab`...

_single cursor (visual mode)_ and _single "phantom" cursor (no mode)_
<img width="270" alt="Screen Shot 2022-02-01 at 6 48 11 PM" src="https://user-images.githubusercontent.com/14429506/152075577-e4b8a928-b651-48dd-a2fe-aab488e00215.png">

(you would expect there to only be a single cursor in insert mode) 

```
vscodevim v1.21.10
tabnine v3.5.18
```

**Comment by LeeroyDing** [comment]
TabNine is also affected by https://github.com/yzhang-gh/vscode-markdown/issues/1011.

The workaround mentioned in https://github.com/yzhang-gh/vscode-markdown/issues/1011 doesn't fully work. The only way to fix it is to add this into `keybindings.json`:

```
  {
    "key": "tab",
    "command": "-markdown.extension.onTabKey",
    "when": "editorTextFocus && !editorReadonly && !editorTabMovesFocus && !hasOtherSuggestions && !hasSnippetCompletions && !inSnippetMode && !suggestWidgetVisible && editorLangId == 'markdown'"
  }
```

This effectively disables the command on `markdown.extension.onTabKey`. I suspect that TabNine inline completion does not modify the value of `inlineSuggestionVisible`, causing this workaround to not work.

**Comment by yonip23** [comment]
Thank you all for the feedback - I'll discuss with my team regarding the customization of the inline keybindings, looks like it'll solve some of the issues brought up here.

**Comment by jacobhtye** [comment]
I love TabNine btw, but the inline completion just tends to get in the way. Tried it for a while thinking I would love it but its just kinda annoyting. Maybe I would liek it if it still showed the multiple suggestions I could choose from, but the first one is not often the one I want

**Comment by T3kla** [comment]
I was trying to accept inline completions but it just added a tab. Seems like it was an incompatibility with TabOut because it got fixed when I disabled TabOut.

**Comment by Dark-Wolf60** [comment]
I loved inline completions but it doesn't working anymore and the suggestions are popping up. I want just the inline completions.
I don't want the suggestions it interrupts my emmet suggestions. how to fix it?

**Comment by yonip23** [comment]
> I loved inline completions but it doesn't working anymore and the suggestions are popping up. I want just the inline completions. I don't want the suggestions it interrupts my emmet suggestions. how to fix it?

Hey @Dark-Wolf60, we'll be happy to help you over at support@tabnine.com 🙂

**Comment by realKarthikNair** [comment]
I'm exactly facing the same.
> I loved inline completions but it doesn't working anymore and the suggestions are popping up. I want just the inline completions. I don't want the suggestions it interrupts my emmet suggestions. how to fix it?

**Comment by davidyeong97** [comment]
I only use tabnine for its inline completions as it doesn't interfere with other emmet suggestions, but it stopped working. I tried to revert to older updates but still no luck. Any suggestions on how to fix it?

**Comment by realKarthikNair** [comment]
@davidyeong97  Same thing here... And no one seems to give a damn. Im on linux btw

**Comment by yonip23** [comment]
@davidyeong97 @realKarthikNair y like dis, we do give a damn 🙂 
we are working on a fix, should be available soon enough

**Comment by realKarthikNair** [comment]
@yonip23  I'm sorry, i was going through a tough time and seeing tabnine broken added fuel to the fire provided i was a premium user so I typed so

**Comment by yonip23** [comment]
@realKarthikNair we all have our ups and downs, dont sweat it :)
![image](https://user-images.githubusercontent.com/67855609/157281000-5fa56183-c13c-481a-a756-70947bae5860.png)

**Comment by realKarthikNair** [comment]
@yonip23  Thanks and Sorry

**Comment by yonip23** [comment]
@realKarthikNair Thank you, I believe we will release a binary version with a fix soon enough - will update here once the version is available

**Comment by finnmerlett** [comment]
Hi there @IsaacShelton and all,

I saw what you wrote here:
> Cool feature, but it seems to break when using with VSCodeVim. Pressing `tab` to accept a completion does the completion successfully, but the caret in split in two, with one visual mode cursor and one weird cursor that isn't in any mode.
> 
> Have no idea what why
> 
> Typing code...
> 
> _single cursor (insert mode)_ <img alt="Screen Shot 2022-02-01 at 6 48 04 PM" width="270" src="https://user-images.githubusercontent.com/14429506/152075578-5c9c5818-a600-4a19-9b6b-f7ec5283d20c.png">
> 
> After `tab`...
> 
> _single cursor (visual mode)_ and _single "phantom" cursor (no mode)_ <img alt="Screen Shot 2022-02-01 at 6 48 11 PM" width="270" src="https://user-images.githubusercontent.com/14429506/152075577-e4b8a928-b651-48dd-a2fe-aab488e00215.png">
> 
> (you would expect there to only be a single cursor in insert mode)
> 
> ```
> vscodevim v1.21.10
> tabnine v3.5.18
> ```

I'm having exactly this issue, using VSCodeVim with Tabnine, with inline suggestions (what a great feature idea btw) switched on. Looking forward to having a proper fix - in the meantime, I found a multi-command workaround. Simply add the `multi-command` extension by ryuta46 to VSCode, then add these to the bottom of your keyboard shortcuts file.

```json
  {
    "key": "tab",
    "command": "-tabnine.accept-inline-suggestion",
    "when": "tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
  },
  {
    "key": "tab",
    "command": "extension.multiCommand.execute",
    "args": {
      "sequence": [
        "tabnine.accept-inline-suggestion",
        "extension.vim_escape",
        "extension.vim_escape",
        "extension.vim_insert",
        "cursorEnd",
      ],
      "interval": 20, 
    },
    "when": "tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
  },
```

The first command removes the Tabnine accept inline suggestion `tab` shortcut, the second entry replaces it (with the exact same key and `when` configuration) with a command sequence that accepts the suggestion, then:
1. exits the glitched vim visual mode (esc x2)
2. switches to vim insert mode (ins)
3. moves the cursor to the end of the line

On my computer, the interval had to be 20ms or greater - otherwise the command execution overlapped and messed up the sequence.

**Comment by yonip23** [comment]
@frizzant Thank you for sharing! we are planning on enabling to customize keyboard shortcuts for inline completions, but until then its nice to have this workaround.

**Comment by yonip23** [comment]
@realKarthikNair and everyone experiencing issues with inline completions not working - a new engine version (4.1.16) is released to beta with a fix. If you're a beta user - We'd like to know if this fix is working for you guys. 
Thank you all!

---

- To enable beta, go to the `Hub -> Preferences -> Receive beta releases of Tabnine`
- To make sure you're running the correct version, make sure that `Hub -> Preferences -> Engine Version` says `4.1.16-xx`

**Comment by finnmerlett** [comment]
Checked and I'm on the beta (4.1.16-41) - the issue still seems to be there when I switch off my shortcut workaround. NOTE: it only errors when you've typed one/a few letters and auto-complete when your cursor is in the middle of the new would-be word. If you press spacebar then autocomplete, the issue doesn't occur.

**Comment by yonip23** [comment]
@finnmerlett this version is tackling another issue, unfortunately not yours 😞 
There was this issue where users with inline completions enabled would get popup completions instead.

**Comment by realKarthikNair** [comment]
@yonip23 Thanks it works now. Lots of love.. I can finally die in peace

**Comment by yonip23** [comment]
@realKarthikNair no need to die, you can just code in peace now 😃

**Comment by realKarthikNair** [comment]
Thank you!!

**Comment by yonip23** [comment]
@finnmerlett @veeyslaw and all,
### Regarding Customizing keyboard shortcuts for inline completions
I checked and was able to customize my keyboard shortcuts in this manner:
### Vscode:
According to [this guide](https://code.visualstudio.com/docs/getstarted/keybindings#_customizing-shortcuts), open this view:
![image](https://user-images.githubusercontent.com/67855609/157726630-8607b309-5520-4473-a4ba-3d30487f2c98.png)
Search for "tabnine":
![image](https://user-images.githubusercontent.com/67855609/157726737-ad26e4fe-89b0-4313-85dd-821383f880d5.png)
and edit to whatever shortcut you desire.

### JetBrains
According to [this guide](https://www.jetbrains.com/help/idea/configuring-keyboard-and-mouse-shortcuts.html), open `Settings -> Keymap -> Plugins -> Tabnine AI Code Completion- JS Java Python TS Rust Go PHP & More`
![image](https://user-images.githubusercontent.com/67855609/157727640-9ce67718-c23c-45b2-a569-23d65e1f3e0a.png)
 Double click the desired action -> `Add Keyboard Shortcut`
![image](https://user-images.githubusercontent.com/67855609/157727883-6fd5cf47-7338-454a-a89f-27afbdf59634.png)
And set it to whatever you want.

---

Please try this solution and let me know if its what you guys were looking for, and if not what's missing.

Kind regards

**Comment by finnmerlett** [comment]
@yonip23 thank you for looking into this! Unfortunately, that the issue here is nothing to do with the keybinding (in fact, my previous comment already showed I was aware of how to change the keybinding for accepting the Tabnine inline suggestion!). To confirm, I tried binding the accept inline suggestion command to a random keybinding and tested it, and got the same ghost cursor and incorrect behaviour.

This issue isn't to do with keybinding, it is to do with interaction with VSCodeVim's various cursor modes - they don't play well together, and somehow the Tabnine accept inline suggestion code action messes it up causing VSCodeVim to enter the described glitched cursor mode state. It's important to note that normal suggestion acceptance doesn't cause this glitch, so it's definitely possible to do correctly/

**Comment by finnmerlett** [comment]
Okay, I have a way simpler workaround! Just disable VSCodeVim before accepting the Tabnine suggestion, then quickly re-enable it!! Simply add the following to your keybindings.json file 😄 

```json

  {
    "key": "tab",
    "command": "-tabnine.accept-inline-suggestion",
    "when": "tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
  },
  {
    "key": "tab",
    "command": "extension.multiCommand.execute",
    "args": {
      "sequence": [
        "toggleVim",
        "tabnine.accept-inline-suggestion",
        "toggleVim",
        "extension.vim_insert",
      ],
      "interval": 50, 
    },
    "when": "vim.active && tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || vim.active && tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
  },
```

**Comment by ThaDaVos** [comment]
I noticed random lines disappearing from my code at random while working, after disabling this feature it stopped - any idea why this is happening all of the sudden?

I am using VSCode btw - maybe it's clashing with https://code.visualstudio.com/updates/v1_60#_inline-suggestions-in-autocomplete ?

**Comment by ThaDaVos** [comment]
I doesn't only remove a single character, also whole lines - it sometimes even happens after clicking around inside my code

@FabKaiz - I'll try your suggestion, I also turned off the empty line completions and I don't see the bug happen that often anymore (changed the coloring of the inline hint too to better check what inline completion is causing this - as VsCode also has it's own inline completion

**Comment by rcoundon** [comment]
In Webstorm I'm repeatedly seeing this error and not sure what the sequence of events is that leads to it, unfortunately

```
Tabnine is not able to function properly. Contact support@tabnine.com

com.tabnine.binary.exceptions.BinaryCannotRecoverException
	at com.tabnine.binary.BinaryProcessRequesterProvider.onDead(BinaryProcessRequesterProvider.java:69)
	at com.tabnine.binary.BinaryRequestFacade.executeRequest(BinaryRequestFacade.java:44)
	at com.tabnine.prediction.CompletionFacade.retrieveCompletions(CompletionFacade.java:77)
	at com.tabnine.prediction.CompletionFacade.lambda$retrieveCompletions$0(CompletionFacade.java:34)
	at com.intellij.openapi.application.ex.ApplicationUtil.lambda$runWithCheckCanceled$1(ApplicationUtil.java:43)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:624)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:698)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:646)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:623)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:66)
	at com.intellij.openapi.application.ex.ApplicationUtil.lambda$runWithCheckCanceled$2(ApplicationUtil.java:41)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:263)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:829)

```

**Comment by letronje** [comment]
How does one accept inline completions in goland ? pressing tab doesn't seem to work.

**Comment by artkad** [comment]
I am using vim plugin with tabnine at vscode and when I choose option by pushing 'tab' vim plugin switched to 'visual' mode and I cant spelling right now. I have to push 'escape' then switch to 'insert' mode and continue coding. It is angry me)

**Comment by sebastianjung** [comment]
@artkad Please look 5 comments above for a workaround to this problem.

**Comment by finnmerlett** [comment]
@artkad In case counting to 5 is too complicated https://github.com/codota/TabNine/issues/432#issuecomment-1064936875

**Comment by Trapfether** [comment]
This is purely based on my experience. I happened to get a new computer since this switch was defaulted to on, so I started using the feature without understanding what was going on. Here are some usability concerns I noticed.

Tabnine suggestions don't show up in the auto-complete popup at all when this is on, meaning only the "Best" suggestion is available, which is incorrect on multiple occasions. It would be better if the suggestions ALSO appear in the popup if you still have it enabled.

Inline suggestions take precedence over auto-complete selections. Even if I arrow down to deliberately select a different auto-complete, hitting tab will accept and input the inline-suggestion. If an arrow key is hit, then the inline suggestion should either be disabled, or changed to reflect the new selected suggestion.

These two changes would in my view make this feature more convenient, less error-prone, and more user-friendly.

**Comment by Staninna** [comment]
@Trapfether i changed my `tabnine.accept-inline-suggestion` to the right key and that fixes your problem number 2 it still doesn't fix problem number 1 tho

to change it go to the command palette and type `Open keyboard shortcuts` press enter in the search bar type `"tab"` and change the shortcut for `tabnine.accept-inline-suggestion` to right arrow by clicking on the pen

I hope this helps

**Comment by Trapfether** [comment]
@Staninna I disabled the feature for now. I'll be looking forward to updates in the future.

**Comment by david-morris** [comment]
I have other completion sources, and this violates the normal way that completion happens. With my (VSpaceMacs) setup, it's unclear how to decline a suggestion, and when I open the Settings for the app, I don't immediately see a toggle to turn this behavior off.

I understand that this behavior is obviously desirable to whoever wrote it as well as a lot of people in this thread, but it's disruptive and should be opt-in.

Not opt-out, because I don't want to worry about where my settings for this option are saved.  Default behavior should play nicely with normal operation of VSCode.

**Comment by adanylkov** [comment]
> Okay, I have a way simpler workaround! Just disable VSCodeVim before accepting the Tabnine suggestion, then quickly re-enable it!! Simply add the following to your keybindings.json file smile . This also solves the issue of the cursor always ending up at the end of the line you're working on.
> 
> ```json
>   {
>     "key": "tab",
>     "command": "-tabnine.accept-inline-suggestion",
>     "when": "tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
>   },
>   {
>     "key": "tab",
>     "command": "extension.multiCommand.execute",
>     "args": {
>       "sequence": [
>         "toggleVim",
>         "tabnine.accept-inline-suggestion",
>         "toggleVim",
>         "extension.vim_insert",
>       ],
>       "interval": 50, 
>     },
>     "when": "vim.active && tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || vim.active && tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
>   },
> ```

thank you!

**Comment by allendred** [comment]
This seems to conflict with `copliot`

**Comment by david-morris** [comment]
Hi, I just would like to explain via a workflow anecdote why I think it's important that this feature be opt-in and configurable in the vscode settings.

I've just had to reload window twice in 5 minutes, and was taken out of a flow state once more.

I did a rebuild container without cache and kept working.  All of a sudden, my typing is randomly interrupted by a wild RNN in overtype mode, replacing `{4} />` with `{404: File not found`.  I wonder what addon did this?  I've already disabled inline completions, so it can't be tabnine.  Well, I'll check anyway, and... I've been signed out of tabnine.  OK, I'll just sign in and get back to work.

I relax into my code again, and then, bam! Something else gets overwritten, breaking syntax.  I remember now, when you change the inline completions setting, you have to reload the window.  So I do that and get back to work.

Glad to have that behind me, I start writing again, and... something else gets overwritten.  I revert the file, wondering if some other addon is causing this, and open the tabnine hub.  Hmm, my settings were changed to accept inline suggestions!  It must have changed my account setting to the default when I logged in after implicitly loading defaults by rebuilding the container without the cache.  Which is also a bug.

OK, so I set my account setting back to the only value I've ever voluntarily supplied, and reopen the window because that's needed to change that setting, and I wait for Tabnine to load and check the settings before I fully relax.

I'm a little bit upset.

**Comment by LeshemNoa** [comment]
Hello all, regarding the Vim for VSCode and Tabnine inline completions discrepancy.
We just released a fix for the Vim mode switching issue and the incorrect cursor positioning issue, that occur when accepting an inline completion from Tabnine. The fix is available with version 3.5.49 of Tabnine's VSCode plugin. We'd love for you to try it out and let us know what you think. 

Big thanks to everybody who raised this issue!
@finnmerlett @IsaacShelton @hugohabicht01 @shadowndacorner @adanylkov @xuzhanhh @lastgosumatt @AliXEagle

**Comment by PixelsDE** [comment]
Well. Its cool to have a advanced completion feature but its sad that its only possible with "inline" feature

**Comment by david-morris** [comment]
I've noticed that if you set the inline hint color to "none" in the settings, that seems to override having turned off the inline completions in the settings page.  That's a bug.

**Comment by david-morris** [comment]
One thing about when TabNine ignores the preference to not give inline suggestions is that if the setting is set off both locally and in the cloud, when you rebuild a dev container, it will give inline suggestions anyway until you toggle the setting on and off in the settings page.
So there seems to be something wrong with some code somewhere that synchronizes preferences.  Maybe there was an assumption about when preferences needed to be read which is violated by rebuilding a container?

**Comment by kouhe3** [comment]
![image](https://user-images.githubusercontent.com/25522053/173860484-6555aae6-180a-40b4-a52e-189d9434d3a1.png)

both inline + popup completion at the same time?

how to fix this bug

**Comment by skwaugh** [comment]
feedback - I'd like to use Advanced Completions *without* inline completions.   Suggest the Advanced completion first within the popup, but don't make me give up visibility of other suggestions within the Popup

Also - Under Advanced Completions, there is a `Note: This also enables cloud mode`.   
You should indicate that it also enables/disables Inline Completions, and that you can't have one without the either (or change that :-] )

Thanks!

**Comment by jabbalaci** [comment]
I use ZSH as my shell and there the inline completion is the right arrow. I'm used to it and in VS Code I always press it instead of TAB, but it doesn't work as expected.

I tried to add it to `keybindings.json`:
```
    {
        "key": "right",
        "command": "tabnine.accept-inline-suggestion",
        "when": "tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
    },
```

However, no effect. I want the right arrow to function just like the TAB for completing inline suggestions. Thanks.

**Comment by Carleslc** [comment]
I can't see more than one suggestion using inline completions. In my case, the `alt +[` and `alt +]` keys are not working as stated in some comments above.

I'm using VSCode plugin version 3.5.58. I don't see either the `tabnine.next-inline-suggestion` or `tabnine.prev-inline-suggestion` in the keyboard shortcuts settings.

Anyone can provide some help on this? Thank you.

**Comment by chirag127** [comment]
I can't know if the suggestions are made by GitHub copilot or tab nine when this is enabled. I just enabled it and started to think why the hell GitHub copilot got bad and then found its tab nine.

**Comment by laurentpayot** [comment]
When editing a comment in VSCode, the proposed words are the same color as the comment. It makes it impossible to distinguish  the actual completion from the end of the commented line. It is such a pain that I had to disable inline completions.
IMHO a color picker for the inline completions should be added in preferences.

**Comment by saarelad** [comment]
It was probably already mentioned before, but anyway:

When working on intelliJ with live templates configuration, tabnine's inline suggestions overlap the template, see example:
 
![image](https://user-images.githubusercontent.com/20183220/178244587-8c871119-1ac7-4847-a1c4-511821f2497f.png)

**Comment by tomdunhamsf** [comment]
This is now required for Advanced Completions, which for me are noticeably better.  Could we have an option to place suggestions both in the pop-up and as an inline because:

1. In practice I will never use Option+[ to cycle through the suggestions.  It is too slow for me to see the alternatives, when instead I could just keep typing to get a better suggestion.  If the suggestions are in the pop-up, I can see right away if TabNine already has what I want as the second or third option.  By placing in inline as well, I will be able to see the full suggestion of whichever option I currently have selected in the pop-up (or the first in the pop-up if I haven't hit the arrow key yet).
2. Sometime the inline gets jumbled with whatever Intellij wants to write to the side of my cursor, like parameter names.  If the suggestion were also in the pop-up, it would help.

Until Advanced Completions, I just used the pop-up.  This works fine for me, because even if the suggestions are short, if I can quickly select them, it speeds me up.  However, Advanced Completions guess things like what String literal I want to use during testing much better.  Frankly, I never really see a multi line match, nor do I need one.  If you can correctly guess the rest of a line after a few keystrokes that is good enough for me.  So even just allowing Advanced Completion to work with pop-ups would work for me.

I understand that you might want inline for Advanced Completion, so that the programmer can see the full suggestion if it is multiline.  I'd suggest just showing the current suggestion as multi-line for this reason, but also displaying the pop-up so that the user can quickly see if there are useful suggestions to cycle to.  As an added bonus the programmer just continues to use the same keys as the built-in IDE completion uses.

Note that I also have full line completions enabled, if that makes a difference.

**Comment by Creling** [comment]
Well, I need an option to **DISABLE** inline completions.

**Comment by Jakandisio** [comment]
Please how to disable inline completions? really hate it. please remove it and show in the popup box instead as it always was.

**Comment by halfelf** [comment]
How to achieve this ? 

[ ] Whole-line code completions
[ ] Full-function code completions
[x] Natural language to code completions
[x] Learns your coding patterns & style

Or, I want to go back to previous version when I could choose between classic completion and inline.

**Comment by vincentinttsh** [comment]
> Well, I need an option to **DISABLE** inline completions for it breaks the classical workflow that the pop-up gathers all suggestions worth noticing, no matter they come from TabNine, IntelliSense or other extensions.
> 
> Tbh, I scarcely have the extra energy to focus on the gray, indistinct sentences produced by Inline Completion. Not to mention these sentences are covered by the pop-up occasionally.

I found a way to disable inline completions:
1. find the location Tabnine store config( Type `Tabnine::config_dir` in your editor to find your configuration directory.)
2. find a file name `tabnine_config.json` and open it
3. change `inline_suggestions_mode` to false

**Comment by Trapfether** [comment]
Hello Again,

A reiteration of a usability Concern I expressed earlier, and an issue related to transparency on this topic.

Per my previous comment, I disabled inline suggestions due to a usability problem. Namely, inline suggestions do not play well with auto-complete options. If I arrow down to select an option from the autocomplete popup and hit tab to select it, tab-nine hijacks that tab press and puts its suggested inline completion instead. This is now compounded by a flaky inline preview issue. The grey text does not always appear when Tab-Nine intends to place an inline completion so I will select an option from the autocomplete dropdown and a completely different completion will appear seemingly out of nowhere.

I would back other comments I have seen here, that autocomplete options need to be placed in the auto-complete modal and inlines disabled by default. In Particular, it should either be an option; or the default, that inline completions are something to be REQUESTED by the user via a shortcut, otherwise, they should be included in the auto-complete popup options instead.

The amount of time wasted fixing mistakes brought about by the inline completions is absurd. In addition to the above usability issues, Tab-Nine highlights the entire completion after hitting the tab to select it, meaning that if I then hit anything other than an arrow, the suggestion I seemingly just selected is now replaced by the new character. So now the keystrokes necessary to work with inline suggestions are a tab to select, an arrow to unhighlight, backspace to the final correct character, and finish the line manually. As opposed to the previous workflow using the suggestion popup, arrow down the list to a correct useful suggestion, hit the tab to select it and go about your day.

Finally, the transparency issue. Everything I described above was the result of inline suggestions turning on without my informed knowledge. I once again found myself in the position of not understanding WHY my workflow was suddenly different and broken. It is now apparent that this is the result of turning on advanced completions. I did so when prompted by Tab-Nine within my editor; however, nowhere during that procedure was it made apparent that turning on said feature would also force inline suggestions to turn on. This is especially irritating when you consider that I had the feature turned OFF in my settings. I should have been alerted to the conflict and been able to make an informed decision about whether to continue. Then when I began experiencing the previously mentioned issues, I would have at least understood where they were coming from and what course of action would correct them.

Once Again, I have disabled advanced completions until these usability issues are resolved. I find Tab-Nine to be useful when it is contained to the completions pop-up; however, the inline completion feature is simply obtrusive with its current implementation. A full rework or some more granular options are NECESSARY for this to be smoothly integrated into my; and seemingly many others, workflows.

**Comment by maya-tabnine** [comment]
Hi guys, sorry for not responding here!!
Our team is reading all of your comments, and we will take them into consideration. Every feedback or opinion you share here is brought up to the product and R&D teams so they can improve the Tabnine hub and the whole Tabnine experience going forward to make things better for you.

**Comment by oldmansutton** [comment]
I agree with the complaints regarding being forced into inline suggestions without my knowledge.  I've been using Tabnine for a 
 long time now, and have gotten used to the workflow.  Yesterday I was coding so fast, today the inline suggestions are now the default and I can't find a way to turn them off.  My workflow has been crippled during a time when I NEED to be productive, and I'm spending more time for worse suggestions with the new default.  Not happy with Tabnine now at ALL.

**Comment by david-morris** [comment]
> Hi guys, sorry for not responding here!! Our team is reading all of your comments, and we will take them into consideration. Every feedback or opinion you share here is brought up to the product and R&D teams so they can improve the Tabnine hub and the whole Tabnine experience going forward to make things better for you.

I don't understand. Just today, an automatic tabnine update turned inline completions back on. Is tabnine moving away from providing completions to people who want "just another" completion source?

**Comment by jairdeon** [comment]
Good afternoon.
I've been using Tabnine since its launch and I've been using it together since forever, until the launch of other competitors, which I used or use together.

Today, by coincidence, I noticed that Tabnine is clearly fighting with the other autocomplete AIs, and what caught my attention the most is that the "inline" feature is forcibly active.

I also noticed that apparently the suggestions no longer appear in the popups, and only inline, I also verified that, in the available options of the tabnine, there is no longer the option to "disable online", and or you have all the resources (which includes the inline and multiline padding), or basic padding (also inline).

I particularly feel a loss of productivity because of this, although it is something "futile", but it was how I was used to it since forever. New features are always welcome, but I believe that the user should have the chance to choose what he wants, instead of being "forced" to try a new way of doing something that was already done before (but I respect the business rule chosen by the company).

Some points I noticed
- Tabnine's settings menu has been drastically reduced, not being possible to customize program parameters (either you have all the features, or partially (Local/Hybrid/Cloud), and being all, they are all, regardless of whether you want to use them or not)
- Tabnine fights directly with Github Copilot when trying to suggest lines (in the sense that both suggestions appear together), with normally the first "tab" activates one, and then the other
- When both plugins are active, the other one cannot continue the suggestion made previously (copilot suggests a function, and Tabnine suggests an else for example next)
- Github copilot offers (in PHPStorm) the option to display your suggestions after pressing a key combination (so you can choose if you want it to display suggestions by default or only after pressing the keys), and in Tabnine it is not possible to apply a criteria for the inline suggestion to appear (although I prefer the traditional popup).

From the above resources, I know that "it is not the plugin's responsibility" to work in conjunction with another AI, and that it is best to use just one. But in my workflow I was always using Github Copilot for inline suggestion (I feel it's less invasive) and Tabnine for suggesting parts of words, strings and parameters (in the popup part), and in my case I was getting great results using the 2 together. But I feel that currently, I'm being "forced" to have to choose just one, since before I was able to evaluate the use of the 2 together in my work model and I would like to continue that way.

Unfortunately what is bothering me the most is the issue of "forcing" the inline, really in my case, the popup suggestion was the most suitable (for my workflow), the non-display of the suggestions in the popup, and the removal of customizing the parameters in the plugin configuration.

The quotes above are made constructively, as I am a big fan of the tool, and I wish to continue using it for its greater purpose which is to increase productivity, regardless of rivalries or alternatives.

**Comment by BlueHawk77** [comment]
> Good afternoon. I've been using Tabnine since its launch and I've been using it together since forever, until the launch of other competitors, which I used or use together.
> 
> Today, by coincidence, I noticed that Tabnine is clearly fighting with the other autocomplete AIs, and what caught my attention the most is that the "inline" feature is forcibly active.
> 
> I also noticed that apparently the suggestions no longer appear in the popups, and only inline, I also verified that, in the available options of the tabnine, there is no longer the option to "disable online", and or you have all the resources (which includes the inline and multiline padding), or basic padding (also inline).
> 
> I particularly feel a loss of productivity because of this, although it is something "futile", but it was how I was used to it since forever. New features are always welcome, but I believe that the user should have the chance to choose what he wants, instead of being "forced" to try a new way of doing something that was already done before (but I respect the business rule chosen by the company).
> 
> Some points I noticed
> 
> * Tabnine's settings menu has been drastically reduced, not being possible to customize program parameters (either you have all the features, or partially (Local/Hybrid/Cloud), and being all, they are all, regardless of whether you want to use them or not)
> * Tabnine fights directly with Github Copilot when trying to suggest lines (in the sense that both suggestions appear together), with normally the first "tab" activates one, and then the other
> * When both plugins are active, the other one cannot continue the suggestion made previously (copilot suggests a function, and Tabnine suggests an else for example next)
> * Github copilot offers (in PHPStorm) the option to display your suggestions after pressing a key combination (so you can choose if you want it to display suggestions by default or only after pressing the keys), and in Tabnine it is not possible to apply a criteria for the inline suggestion to appear (although I prefer the traditional popup).
> 
> From the above resources, I know that "it is not the plugin's responsibility" to work in conjunction with another AI, and that it is best to use just one. But in my workflow I was always using Github Copilot for inline suggestion (I feel it's less invasive) and Tabnine for suggesting parts of words, strings and parameters (in the popup part), and in my case I was getting great results using the 2 together. But I feel that currently, I'm being "forced" to have to choose just one, since before I was able to evaluate the use of the 2 together in my work model and I would like to continue that way.
> 
> Unfortunately what is bothering me the most is the issue of "forcing" the inline, really in my case, the popup suggestion was the most suitable (for my workflow), the non-display of the suggestions in the popup, and the removal of customizing the parameters in the plugin configuration.
> 
> The quotes above are made constructively, as I am a big fan of the tool, and I wish to continue using it for its greater purpose which is to increase productivity, regardless of rivalries or alternatives.

I totally agree with @jairdeon. We should not be forced to use 'one' tool over 'another' even when one of them may be better than the other if there is a way we can use them together.

**Comment by yonip23** [comment]
Hey all,
After taking notes from all your helpful comments, we released a new way to render completions - **Hybrid mode**.
**This is currently available only for JB clients**, but we plan on releasing a similar feature in VSCode soon.

This feature renders **advanced completions** in inline, and **basic completions** in popup mode.

To enable the feature:
1. Update to the latest plugin and binary versions - plugin 0.7.21, binary 4.4.123 and higher.
2. go to the Hub -> Settings -> Additional Settings and enable "Use Hybrid suggestions rendering mode":
![image](https://user-images.githubusercontent.com/67855609/187066945-416afb5d-69af-4cfe-872a-6753aeb51999.png)

This is how it should look like after you enable it:

https://user-images.githubusercontent.com/67855609/187066983-8efcc093-ce73-4437-b4b2-78390bf022c0.mp4

Thank you all for your valuable feedback, and please let us know how does this work out for you 😃

**Comment by jairdeon** [comment]
> Hey all, After taking notes from all your helpful comments, we released a new way to render completions - **Hybrid mode**. **This is currently available only for JB clients**, but we plan on releasing a similar feature in VSCode soon.
> 
> This feature renders **advanced completions** in inline, and **basic completions** in popup mode.
> 
> To enable the feature:
> 
> 1. Update to the latest plugin and binary versions - plugin 0.7.21, binary 4.4.123 and higher.
> 2. go to the Hub -> Settings -> Additional Settings and enable "Use Hybrid suggestions rendering mode":
>    ![image](https://user-images.githubusercontent.com/67855609/187066945-416afb5d-69af-4cfe-872a-6753aeb51999.png)
> 
> This is how it should look like after you enable it:
> 
>  hybrid-demo.mp4 
> Thank you all for your valuable feedback, and please let us know how does this work out for you 😃

Thank you very much, @yonip23, the feature is working perfectly with this update in JB's IDEs.

**Comment by ooliver1** [comment]
I would appreciate if tabnine didn't popup telling me this is an advanced completion every single time it makes a suggestion and I use it.

**Comment by Akv2021** [comment]
> As an FYI for those who are having issues with markdown (@shmkane for one), the markdown extension has a default keybinding for "tab". Removing that keybinding allowed me to use this correctly.
> 
> @dimacodota As for the Vim issues, it seems like [this issue](https://github.com/Microsoft/vscode/issues/13441) may be the cause, but that definitely isn't certain. That may just be an issue of a conflicting keybinding as well.

Can't seem to make it work. Could you please share which keybinding should be overridden ?

**Comment by Jancera** [comment]
> Okay, I have a way simpler workaround! Just disable VSCodeVim before accepting the Tabnine suggestion, then quickly re-enable it!! Simply add the following to your keybindings.json file 😄 . This also solves the issue of the cursor always ending up at the end of the line you're working on.
> 
> ```json
>   {
>     "key": "tab",
>     "command": "-tabnine.accept-inline-suggestion",
>     "when": "tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
>   },
>   {
>     "key": "tab",
>     "command": "extension.multiCommand.execute",
>     "args": {
>       "sequence": [
>         "toggleVim",
>         "tabnine.accept-inline-suggestion",
>         "toggleVim",
>         "extension.vim_insert",
>       ],
>       "interval": 50, 
>     },
>     "when": "vim.active && tabnine.in-inline-suggestions && tabnine.inline-suggestion:enabled || vim.active && tabnine.in-inline-suggestions && tabnine.snippet-suggestion:enabled"
>   },
> ```

I tried to apply both @adanylkov and @letronje @finnmerlett solutions but they could not solve my problem. I am using **Tabnine 3.6.38** and **vscodevim 1.24.3** and the only solution that worked for me was to remove all **Tab** keybindings related to inline suggestions and suggestions and add the following to my keybindings.json

```
{
    "key": "tab",
    "command": "-tabnine.tab-override",
    "when": "suggestWidgetHasFocusedSuggestion && suggestWidgetVisible && tabnine.tab-override && textInputFocus"
  },
  {
    "key": "tab",
    "command": "extension.multiCommand.execute",
    "args": {
      "sequence": [
        "toggleVim",
        "tabnine.tab-override",
        "toggleVim",
        "extension.vim_insert"
      ],
      "interval": 50
    },
    "when": "suggestWidgetHasFocusedSuggestion && suggestWidgetVisible && tabnine.tab-override && textInputFocus"
  },
```

Hope it helps!

**Comment by Jancera** [comment]
I was having a problem with inline suggestions and vscodevim extension, so I tried the solutions proposed by @adanylkov and @finnmerlett but without results. So what I did was disable all my extensions and let only vscodevim and tabnine and in order to fix the inline suggestion acceptance by pressing Tab I just had to add that to my `keybidings.json`.

```json
{
    "key": "tab",
    "command": "-tabnine.tab-override",
    "when": "suggestWidgetHasFocusedSuggestion && suggestWidgetVisible && tabnine.tab-override && textInputFocus"
  }
```

With that, I came back enabling my extensions and I let some of those disabled, so I do not know if that problem is related to another extension that is messing with the Tab pressing or if it is related to the setting that I added, or both :)

Hope it helps!

**Comment by jonathan-dejong** [comment]
@dimacodota  Let's see if I can bring this issue back to life.. :D 
As a sidenote: you should really reconsider how that stale bot is setup as it seems to do more harm than good to your users.

Following up on multiple stale-bot closed threads such as:
https://github.com/codota/TabNine/issues/618
https://github.com/codota/TabNine/issues/461
https://github.com/codota/TabNine/issues/445
somewhat: https://github.com/codota/TabNine/issues/433

I've been a pro user for years and tabnine is steadily getting much much better. I actually really really like it, and I'm gearing towards introducing it to our whole team. 

Here's my constructive feedback.
Since I started using it, and still to this day, there's one glaring issue I have with it in my every day use. 
That is indentation, or maybe more accurately code standards. 

Ideally, I expect tabnine to provide me with code suggestions (regardless if they're inline, popup, multiline what-have-you) that uses the indentation style I have set in either my linter files, or even in my IDE config. 
If my whole project and the file I have open, and my phpcs.xml and my IDE config all say that I indent using tabs it is *uniquely* frustrating that the one thing that is supposed to be the smartest part about my setup is also the only thing not able to follow this simple rule. 

If I could wish for more than that, it's that tabnine also recognises common linter files such as `.eslintrc`, `.stylelintrc` and `phpcs.xml` in the project and applies those rules to it's suggestions. To me this is the missing puzzle piece in truly making for example comment-to-code something that is jawdropping good and a huge timesaver. 
Because right now, depending on the amount of code produced, I will still have to spend time reformatting and structuring the code to fit our standards, which is much less enjoyable work than actually just writing the code myself in the first place. 
It's literally the part of development that actually *has* been automated for years before AI in IDEs became a thing 😅 

If this is on the roadmap, can we be told something about the progress and timeline? If it's not, can you explain to us why that is? We're not unreasonable people here, we can understand if there's a complexity due to x/y/z. It's more frustrating just not knowing. 

Thank you for a great product!
